frame/tips/src/migrations/unreserve_all_funds.rs

```diff
@@ -0,0 +1,246 @@
+	///
+	/// * `BTreeMap<T::AccountId, T::Balance>`: Map of account IDs to their respective total
+	///   reserved balance by this pallet
+	/// * `frame_support::weights::Weight`: The weight of this operation.
+	fn get_deposits() -> (BTreeMap<T::AccountId, BalanceOf<T, I>>, frame_support::weights::Weight) {
+		let tips = crate::Tips::<T, I>::iter().collect::<Vec<_>>();
+		let tips_len = tips.len();
+		let account_deposits: BTreeMap<T::AccountId, BalanceOf<T, I>> = tips
+			.into_iter()
+			.map(|(_hash, open_tip)| open_tip)
+			.fold(BTreeMap::new(), |mut acc, tip| {
+				// Add the balance to the account's existing deposit in the accumulator
+				*acc.entry(tip.finder).or_insert(Zero::zero()) = acc
+					.entry(tip.finder.clone())
+					.or_insert(Zero::zero())
+					.saturating_add(tip.deposit);
+				acc
+			});
+
+		(account_deposits, RocksDbWeight::get().reads(tips_len as u64))
+	}
+}
+
+impl<T: crate::Config<I> + pallet_treasury::Config<I>, I: 'static> OnRuntimeUpgrade
+	for UnreserveAllFunds<T, I>
+where
+	BalanceOf<T, I>: Sum,
+{
+	/// Gets the actual reserved amount for each account before the migration, performs integrity
+	/// checks and prints some summary information.
+	///
+	/// Steps:
+	/// 1. Gets the deposited balances for each account stored in this pallet.
+	/// 2. Collects actual pre-migration reserved balances for each account.
+	/// 3. Checks the integrity of the deposited balances.
+	/// 4. Prints summary statistics about the state to be migrated.
+	/// 5. Returns the pre-migration actual reserved balance for each account that will
+	/// be part of the migration.
+	///
+	/// Fails with a `TryRuntimeError` if somehow the amount reserved by this pallet is greater than
+	/// the actual total reserved amount for any accounts.
+	#[cfg(feature = "try-runtime")]
+	fn pre_upgrade() -> Result<Vec<u8>, sp_runtime::TryRuntimeError> {
+		use codec::Encode;
+		use frame_support::ensure;
+
+		// Get the Tips pallet view of balances it has reserved
+		let (account_deposits, _) = Self::get_deposits();
+
+		// Get the actual amounts reserved for accounts with open tips
+		let account_reserved_before: BTreeMap<T::AccountId, BalanceOf<T, I>> = account_deposits
+			.keys()
+			.map(|account| (account.clone(), T::Currency::reserved_balance(&account)))
+			.collect();
+
+		// The deposit amount must be less than or equal to the reserved amount.
+		// If it is higher, there is either a bug with the pallet or a bug in the calculation of the
+		// deposit amount.
+		ensure!(
+			account_deposits.iter().all(|(account, deposit)| *deposit <=
+				*account_reserved_before.get(account).unwrap_or(&Zero::zero())),
+			"Deposit amount is greater than reserved amount"
+		);
+
+		// Print some summary stats
+		let total_deposits_to_unreserve =
+			account_deposits.clone().into_values().sum::<BalanceOf<T, I>>();
+		log::info!("Total accounts: {}", account_deposits.keys().count());
+		log::info!("Total amount to unreserve: {:?}", total_deposits_to_unreserve);
```

Comment:
What stats does it report on Kusama and Polkadot?

Reply:
Do you also have them for Polkadot? Just to gauge the values.

Reply:
Based on a snapshot taken ~1 week ago:
```
ElectionsPhragmen total impacted accounts: 323
ElectionsPhragmen stake to unlock: 240,352,972 DOT
ElectionsPhragment deposit to unreserve: 5,668 DOT

Democracy total impacted accounts: 5125
Democracy stake to unlock: 44,792,283 DOT
Democracy deposit to unreserve: 2,308 DOT

Tips total impacted accounts: 12
Tips amount to unreserve: 37 DOT
```

proportionally, a lot less than Kusama